### PR TITLE
Add new rule for clean architecture domain and integration events

### PR DESCRIPTION
**Pull request by william-liebenberg**

### CODE REVIEW

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+redirects:
+---
+
+Domain Events and Integration Events are powerful patterns used in Domain-Driven Design (DDD) to enable loosely coupled communication between different parts of an application. They serve as a means of **notification** for important domain concepts that have occurred.
+
+<!--endintro-->
+
+## What are Domain Events
+
+In the Domain Layer of the [Clean Architecture](https://github.com/sswconsulting/SSW.CleanArchitecture), Domain Events are an integral part of the Domain Model.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:21]: I think this could be a little confusing or even misleading. It's integral to SSW's Clean Architecture template (and solutions), but it's not part of Clean Architecture, as an architectural pattern.

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+Domain Events are raised within the Domain Layer of your application when an entity or aggregate makes a significant decision or undergoes a state change. The main purpose of Domain Events is to enable loose coupling and keep domain logic isolated from the application's infrastructure.
+
+For example, when an `Order` is placed, a domain event can be raised (e.g. `OrderReceivedEvent`) to notify other parts of the domain that need to react to this event, such as updating inventory or sending confirmation emails.
+
+Another example could be when the `Order`'s `Status` changes from `OrderStatus.Received` to `OrderStatus.Processing`, we can publish an `OrderStatusChanged` event.
+
+It's essential to identify suitable events as not everything qualifies as a Domain Event. For instance, "client walked into the store" or "chicken crossed the road" may not be appropriate for a Domain Event.
+
+Domain events are immutable and can be considered as **historical facts** capturing something that occurred in the domain process. They are meant to be a representation of past events and cannot be altered or disputed.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:31]: Awesome ✅

Comment (matt-goldman) [reply rules/when-to-use-domain-and-integration-events/rule.md:31]: Should domain be capitalised for consistency?

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+When describing domain events, we commonly use a past-tense naming convention, such as `OrderCreated`, `UserRegistered`, `InvoiceConsolidated`.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:33]: I wonder if this should be further up. Also wonder if we should break this down into sub-headers (like Naming Domain Events for example).

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+Domain events should not depend on external dependencies or external systems, adhering to the principles of the Domain Layer.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:35]: Although ours depend on MediaR...perhaps we should clarify what we mean by this.

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+When publishing a domain event, the entire Entity or Aggregate can be included since the event's scope is confined to the current bounded context.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:37]: Maybe a link here to the relevant DDD concept?

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+Note: It is important to remember that the definition and behavior of a `Product` in one bounded context might differ from another bounded context, like an e-commerce application product versus the product of a chemical reaction in a laboratory.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:39]: I don't think this note adds anything...could take it out. If we keep it, suggest putting it in an info box.

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+In the `Application` Layer, domain events are typically handled synchronously. Any database side-effects are tracked as part of the current transaction of the original request, ensuring strong consistency in the response sent back to the user once the transaction is committed.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:41]: Didn't we work out in our chat last week that this isn't the approach to use?

> [18/07 5:30 pm] Daniel Mackay [SSW]
> 
> TLDR; should work fine, but be aware of the differences between using Find() vs something like ToListAsync()
> 
> [18/07 5:32 pm] Matt Goldman [SSW]
> 
> Also use Saved I think?
> 
> 
> [18/07 5:33 pm] Matt Goldman [SSW]
> 
> So it should publish the events after the save is completed, and any additional changes should be a new query
> 
> 
> [18/07 5:34 pm] William Liebenberg [SSW]
> 
> yea...multiple sequential transactipons
> 
> [18/07 5:36 pm] Daniel Mackay [SSW]
> 
> Thinking about this more, it makes sense.  
> 
>  * Command updates domain
>  * Events fired
>  * Event Handlers triggered
>  * More domain updates
>  * More events fired (possibly)
>  * More event handlers triggered (possibly)
>  * (and so on and so forth)
> 
>  
> 
> [18/07 5:37 pm] Daniel Mackay [SSW]
> 
> And watch out for any infinite loops!


I think this probably needs its own rule. And I'm sure @danielmackay is keen to write `Rules to better DDD` anyway :)

Comment (matt-goldman) [reply rules/when-to-use-domain-and-integration-events/rule.md:41]: Or...maybe I keep reading and the process in that chat is for integration events...?

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+## What are Integration Events
+
+Integration Events are used for communication between different bounded contexts (or microservices in a distributed system) and enable potentially long-running asynchronous operations like sending large number of emails, generating thumbnail images, or performing additional business logic.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:45]: change:
> sending large number of emails

To:
sending **a** large number of emails

Or:
sending large number**s** of emails

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+In terms of Clean Architecture, Integration Events can be published from both the `Domain` and `Application` layers depending on the requirements of the applications.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:47]: requirements of the application~~s~~

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+Integration Events can be raised from the `Application` layer when the need for communication or coordination between different parts of the application arises. For example, after a specific use case (command / query) is handled successfully, the `Application` layer might raise an Integration Event to notify other microservices or external systems about the outcome of that use case.
+
+These events are usually published **after** the original transaction completes and are typically dispatched through a Message Broker or Event Bus (e.g., Azure Service Bus, RabbitMQ, Redis PubSub). To ensure reliability and consistency, systems often implement mechanisms like a Transactional Outbox.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:51]: A few links would be useful here, eg to the different busses, and to something explaining the Transactional Outbox pattern.

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+Ultimately, the decision on whether to raise Integration Events from the `Application` layer or the `Domain` layer depends on the specific design and requirements of the application. However, it is essential to ensure that raising Integration Events from the `Domain` layer doesn't lead to tight coupling between the domain logic and the infrastructure concerns. If raising Integration Events from the `Domain` layer is chosen, it's good practice to use an abstraction to decouple the domain entities from the actual event handling mechanisms, ensuring a clean separation of concerns.
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:53]: This paragraph is great. It sounds like we are leaning toward making a recommendation to raise Integration Events from the Application layer, is that right? If so, we should just make that recommendation I think, along with the caveat that there may be circumstances where you choose to raise it from Domain for specific reasons. And if you do, heed the warning already here.

If I've misread that, perhaps some guidance on when to use each approach?

File: rules/when-to-use-domain-and-integration-events/rule.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+## When to use Domain vs Integration Events
+
+Use Domain Events within the `Domain` layer to decouple domain-specific logic and enable better maintainability and testability within the core of your application.
+
+Use Integration Events (preferably from the `Application` layer) when you need to communicate and coordinate between different parts of an application or distributed system (such as microservices), to achieve eventual consistency and loose coupling among services.
+
+For an example implementation of Domain and Integration events, check out this [example project](https://github.com/william-liebenberg/CleanArchitectureWithDomainEvents).
```
Comment (matt-goldman) [inline rules/when-to-use-domain-and-integration-events/rule.md:61]: This is really cool...would be nice if this was mentioned further up, and you could link to the relevant parts throughout the rule.

**Review by matt-goldman — APPROVED** [review]
LGTM - I've added some comments for consideration - take/leave :)

File: rules/when-to-use-domain-and-integration-events/rule.md (hunk 1 above)
Comment (william-liebenberg) [reply rules/when-to-use-domain-and-integration-events/rule.md:21]: I'll make it clear that we are talking about CA + DDD to strengthen this idea.

File: rules/when-to-use-domain-and-integration-events/rule.md (hunk 6 above)
Comment (william-liebenberg) [reply rules/when-to-use-domain-and-integration-events/rule.md:39]: Think of, we only return DTOs from the commands/queries... we don't return entities. Same idea that the domain events that contain whole entities don't go beyond the app/infra layer since it might not be well understood outside of the application.

But yes, i think an info box would look nicer :)

File: rules/when-to-use-domain-and-integration-events/rule.md (hunk 7 above)
Comment (william-liebenberg) [reply rules/when-to-use-domain-and-integration-events/rule.md:41]: yea the Domain event side-effects can occur is in the original transaction (Saving) like we have it now. Integration events should happen after the transaction is committed and asynchronously as well so that we don't block the application flow/user request. I'll have these examples in the demo app.

Comment (christoment) [inline rules/when-to-use-domain-and-integration-events/rule.md:41]: I think "Synchronous" can mislead people thinking about async here, but I think this is the best term for it, I couldn't find any better word 😔

**Review by christoment — APPROVED** [review]
Looks good!

File: rules/when-to-use-domain-and-integration-events/rule.md (hunk 11 above)
Comment (william-liebenberg) [reply rules/when-to-use-domain-and-integration-events/rule.md:53]: Yea i had another whole paragraph about this but it ended up a bit waffley... I'll see if i can rework this to highlight the recommendation to raise them from different layers and not the same layer. But it really does depend on the domain... sometimes you just have a requirement that a notification MUST be an email, but we can always do our best to abstract these things out and remove the infra dependencies from domain.

File: rules/when-to-use-domain-and-integration-events/rule.md (hunk 7 above)
Comment (william-liebenberg) [reply rules/when-to-use-domain-and-integration-events/rule.md:41]: ah yea, i see what you mean... i guess i should say `in-process` for Domain events rather than `async background` for Integration events

File: rules/when-to-use-domain-and-integration-events/rule.md (hunk 3 above)
Comment (william-liebenberg) [reply rules/when-to-use-domain-and-integration-events/rule.md:33]: not sure i can move it up higher, i have a few points that build up towards this part and doesn't seem to fit if i move this higher up.

Comment (william-liebenberg) [reply rules/when-to-use-domain-and-integration-events/rule.md:33]: actually... after a bit more fuffing around, i made it work :)

File: rules/when-to-use-domain-and-integration-events/rule.md (hunk 5 above)
Comment (william-liebenberg) [reply rules/when-to-use-domain-and-integration-events/rule.md:37]: Which concept? Bounded Context?
